notes/conds.md

```diff
@@ -0,0 +1,86 @@
+
+### Operators
+
+| Operator     | Function Signature                | Description                                | Example                                        |
+| ------------ | --------------------------------- | ------------------------------------------ | ---------------------------------------------- |
+| `/${string}` |`[] -> Any`                        | Selects value using JSON Pointer            | `["/foo/0"]`                                  |
+|              |`[] -> Any`                        | Selects whole value like "" in JSON Pointer | `[]`                                          |
+| `==`         |`[Exp<Any>, Exp<Any>] -> Bool`     | Equality                                    | `["==", ["/size"], 1]`                        |
+| `<`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than                                   | `["<",  ["/size"], 1024]`                     |
+| `<=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than or equal                          | `["<=", ["/size"], 1024]`                     |
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+-->
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+     ["some", "?x"],
+     [".foo..bar.[].baz", "$", "?x"],
+     ["match", "?x", "*@example.com"],
+      
+     [".foo.quux", "$", "?y"],
+     [">", "?y", 42],
+     ["<", "?x", "?y"]
+  ]
```

Comment:
I have put together prototype implementation of this https://observablehq.com/@gozala/ucan-policy-language although I replaced `$` with `?` to keep things simpler

Reply:
Also instead of `["some", "?x"]` I have `["in", "?member", "?group"]` which made more sense to me but gives you the same effect.

Reply:
I attempted to introduce `["of", "?member", "?group"]` although maybe `for` would have been a better name, with an intent to give it `every` semantics. Version in observable is broken, because it attempt to unify each member of the group which is not going to work.

I have got an idea that I think may work however. I think I can introduce `Member` to represent all members of the group, and during constraints / predicates application only forward if all the members satisfy it.

I'm not exactly sure about unification, if both are `Member` they probably should have same set of members, and if only one is `Member` that all of it's members should unify with the other.

I'll try and report back

Reply:
Moved to https://github.com/ucan-wg/delegation/issues/8